File: frontend/src/lib/derivados.ts
/**
 * Indicadores derivados: densidad poblacional y tiempo de duplicación.
 *
 *   densidad_hab_km2         = población total / extensión territorial (km²)
 *   tiempo_duplicacion_anios = 70 / tasa de crecimiento anual (%)   [regla del 70]
 *
 * Los dos son cocientes de otros indicadores, así que se calculan en vez de
 * depender de que el documento fuente los trajera. La misma política vive en
 * `backend/app/seed/derivados.py`, que la aplica al generar los JSON; esto la
 * repite sobre lo que llega por API para que valga en todas las vistas —tabla,
 * mapa, ficha, gráficas, panel y exportación— incluida la fila de total nacional,
 * que no sale de ningún documento.
 *
 * Política:
 *  1. Falta el valor y hay insumos → se calcula.
 *  2. El valor existe pero contradice a la fórmula por más de 1.5x → gana la fórmula.
 *  3. No hay insumos → se conserva el valor publicado.
 */
import type { Departamento, Indicadores } from "@/types/departamento";
import type { Municipio } from "@/types/municipio";

/** Regla del 70: con r = 2.00 % anual, 70 / 2.00 = 35 años. */
export const CONSTANTE_DUPLICACION = 70;

const TOLERANCIA = 0.5;

export function calcularDensidad(
  poblacion: number | null | undefined,
  superficie: number | null | undefined
): number | null {
  if (!poblacion || !superficie || superficie <= 0) return null;
  return Math.round((poblacion / superficie) * 100) / 100;
}

export function calcularDuplicacion(tasa: number | null | undefined): number | null {
  if (!tasa || tasa <= 0) return null;
  return Math.round((CONSTANTE_DUPLICACION / tasa) * 100) / 100;
}

function resolver(publicado: number | null, calculado: number | null): number | null {
  if (calculado === null) return publicado;
  if (publicado === null) return calculado;
  const mayor = Math.max(publicado, calculado);
  const menor = Math.min(publicado, calculado);
  return menor <= 0 || mayor / menor > 1 + TOLERANCIA ? calculado : publicado;
}

const num = (v: unknown): number | null => (typeof v === "number" ? v : null);

/** Aplica ambas fórmulas a un registro con campos de indicador. */
export function completarDerivados<T extends Record<string, unknown>>(
  registro: T,
  superficie: number | null
): T {
  const densidad = resolver(
    num(registro.densidad_hab_km2),
    calcularDensidad(num(registro.poblacion_total), superficie)
  );
  const duplicacion = resolver(
    num(registro.tiempo_duplicacion_anios),
    calcularDuplicacion(num(registro.crecimiento_anual_pct))
  );
  if (
    densidad === num(registro.densidad_hab_km2) &&
    duplicacion === num(registro.tiempo_duplicacion_anios)
  ) {
    return registro;
  }
  return { ...registro, densidad_hab_km2: densidad, tiempo_duplicacion_anios: duplicacion };
}

/** La extensión del departamento vive fuera del corte anual, no en los indicadores. */
export function completarDepartamento<T extends Departamento>(depto: T): T {
  if (!depto?.indicadores) return depto;
  const indicadores = completarDerivados(
    depto.indicadores as unknown as Record<string, unknown>,
    depto.superficie_km2
  ) as unknown as Indicadores;
  return indicadores === depto.indicadores ? depto : { ...depto, indicadores };
}

export function completarMunicipio(muni: Municipio): Municipio {
  if (!muni) return muni;
  return completarDerivados(
    muni as unknown as Record<string, unknown>,
    muni.superficie_km2
  ) as unknown as Municipio;
}

File: backend/app/seed/derivados.py
"""Indicadores derivados: densidad poblacional y tiempo de duplicación.

Ambos son cocientes de otros dos indicadores, así que se calculan en vez de
copiarse cuando el documento fuente no los trae:

    densidad_hab_km2         = población total / extensión territorial (km²)
    tiempo_duplicacion_anios = 70 / tasa de crecimiento anual (%)   [regla del 70]

Política, idéntica en los dos scripts de extracción y en el frontend
(`frontend/src/lib/derivados.ts`, que la repite para los datos que llegan por API):

1. Si falta el valor y se pueden aplicar los insumos, se calcula.
2. Si el valor existe pero contradice a la fórmula por más de `TOLERANCIA`, gana la
   fórmula: significa que el número publicado quedó obsoleto (por ejemplo, una
   densidad que venía con una población anterior a la que ahora tenemos) o que es un
   error del documento.
3. Si faltan los insumos, se conserva el valor publicado.
"""
from __future__ import annotations

# Regla del 70: una población que crece r % al año duplica su tamaño en 70/r años.
# Ej.: r = 2.00 → 70 / 2.00 = 35 años.
CONSTANTE_DUPLICACION = 70.0

# Desacuerdo tolerado entre el valor publicado y el calculado: hasta 1.5x en
# cualquier dirección. Más que eso ya no es redondeo ni un insumo ligeramente
# distinto, es un dato equivocado.
TOLERANCIA = 0.5


def densidad(poblacion: float | None, superficie: float | None) -> float | None:
    if not poblacion or not superficie or superficie <= 0:
        return None
    return round(poblacion / superficie, 2)


def tiempo_duplicacion(tasa_crecimiento: float | None) -> float | None:
    if not tasa_crecimiento or tasa_crecimiento <= 0:
        return None
    return round(CONSTANTE_DUPLICACION / tasa_crecimiento, 2)


def _resolver(publicado: float | None, calculado: float | None) -> tuple[float | None, str]:
    """Devuelve (valor, motivo) donde motivo ∈ {"publicado", "calculado", "corregido"}."""
    if calculado is None:
        return publicado, "publicado"
    if publicado is None:
        return calculado, "calculado"
    mayor, menor = max(publicado, calculado), min(publicado, calculado)
    if menor <= 0 or mayor / menor > 1 + TOLERANCIA:
        return calculado, "corregido"
    return publicado, "publicado"


def completar(registro: dict, superficie: float | None = None) -> dict[str, int]:
    """Rellena en sitio densidad y tiempo de duplicación. Devuelve el conteo por motivo.

    `superficie` permite pasar la extensión desde fuera (los indicadores de
    departamento no la llevan: vive en el departamento, no en el corte anual).
    """
    conteo = {"calculado": 0, "corregido": 0}
    sup = superficie if superficie is not None else registro.get("superficie_km2")

    for clave, calculado in (
        ("densidad_hab_km2", densidad(registro.get("poblacion_total"), sup)),
        ("tiempo_duplicacion_anios", tiempo_duplicacion(registro.get("crecimiento_anual_pct"))),
    ):
        valor, motivo = _resolver(registro.get(clave), calculado)
        registro[clave] = valor
        if motivo in conteo:
            conteo[motivo] += 1

    return conteo
